contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,450 @@
+ ```
+ that underneath will have a completely separate implementation.
+
+ In this design, we will focus only on bulk list (get) and watch operations.
+Later, we would like to introduce new resources to support bulk create, update
+and delete operations, but that's not part of this design.
+
+ We will start with introducing `list` resource and supporting the following
+two operation:
+```
+POST /apis/bulk.k8s.io/v1/list <body defines filtering>
+POST /apis/bulk.k8s.io/v1/list?watch=1 <body defines filtering>
+```
+ We can't simply make this an http GET request, due to limitations of GET for
+the size (length) of the url (in which we would have to pass filter options).
+This makes the API look similar to our regular API.
+ Note that once we have list fully implemented, adding this kind of watch will
+be faily simple.
+
+ The main drawback of this approach is that it won't allow for dynamic altering
+of watch subscriptions (which we would like to also support).
+As a result, we need another API for watch that will also support incremental
+subscribtions. That's why we will also introduce the following operation:
+```
+websocket /apis/bulk.k8s.io/v1/list?watch=1
+```
+ The more detailed description of this protocol is in the followin subsection.
+
+*TODO: For consistency, we also considered introducing websocket API for
+handling LIST requests, where first client sends a filter definition over the
+channel and then server sends back the response, but we dropped this for now.
+
+
+### Filtering definition
+
+ We need to make our filtering mechanism to support different resource types at
+the same time. On the other hand, we would like it to be as consistent with all
+other Kubernetes APIs as possible. So we define the selector for bulk operations
+as following:
+
+```
+type BulkListOptions struct {
+	Selectors []ListSelector
+}
+
+type ListSelector struct {
+	Resource GroupVersionResource
+
+	// We would like to reuse the same ListOptions definition as we are using
+	// in regular APIs. For the purpose of bulk operations, you can just specify
+	// more than one.
+	//
+	// NOTE: In the initial implementation, we will require all Options (for a
+	// given resource type) to have exactly the same ResourceVersion - otherwise
+	// we will reject it at the validation level.
+	// This will simplify the initial implementation as we need to guarantee that
+	// watch events within a single resource types are delivered with increasing
+	// resource versions, which may be more tricky with different resource versions.
+	Options []ListOptions
+}
+```
+
+ We will create a dedicated admission plugin (or other filter mechanism) that
+will be responsible for detecting whether a given user is allowed to list/watch
+objects requested by a given `BulkdSelector` (the exact mechanism for it is out
+of scope for this document) and either rejecting the whole request or letting it
+go. We are not going to support partial rejections (e.g. you cannot proceed with
```

Comment:
Terminating the whole watch if an authz fails will make auth changes very expensive. Maybe say we might change this in the future? It doesn't seem so bad to add an error object to the response stream for this case.